File: core/src/main/java/com/github/fburato/highwheelmodules/core/analysis/ModuleAnalyser.java
package com.github.fburato.highwheelmodules.core.analysis;

import com.github.fburato.highwheelmodules.core.algorithms.CompoundAccessVisitor;
import com.github.fburato.highwheelmodules.model.modules.Definition;
import com.github.fburato.highwheelmodules.model.modules.ModuleGraphFactory;
import com.github.fburato.highwheelmodules.utils.Pair;
import com.github.fburato.highwheelmodules.model.classpath.AccessVisitor;
import com.github.fburato.highwheelmodules.model.classpath.ClassParser;
import com.github.fburato.highwheelmodules.model.classpath.ClasspathRoot;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class ModuleAnalyser {

    private final ClassParser classParser;
    private final ClasspathRoot root;
    private final Optional<Integer> evidenceLimit;
    private final ModuleGraphFactory factory;

    public ModuleAnalyser(final ClassParser classParser, final ClasspathRoot root, Optional<Integer> evidenceLimit,
            ModuleGraphFactory factory) {
        this.classParser = classParser;
        this.root = root;
        this.evidenceLimit = evidenceLimit;
        this.factory = factory;
    }

    public List<Pair<Definition, AnalyserModel.AnalysisResult>> analyseStrict(final List<Definition> definitions) {
        if (definitions.isEmpty()) {
            return new ArrayList<>();
        } else if (definitions.size() == 1) {
            return Collections.singletonList(Pair.make(definitions.get(0), internalAnalyseStrict(definitions.get(0))));
        } else {
            return internalAnalyseStrict(definitions);
        }
    }

    private List<Pair<Definition, AnalyserModel.AnalysisResult>> internalAnalyseStrict(
            final List<Definition> definitions) {
        final List<Pair<Definition, AnalysisState>> defVisitors = getAnalysisStates(definitions);
        final AccessVisitor visitor = new CompoundAccessVisitor(collectVisitors(defVisitors));
        try {
            classParser.parse(root, visitor);
        } catch (IOException e) {
            throw new AnalyserException(e);
        }

        return defVisitors.stream().map(p -> Pair.make(p.first, StrictAnalyser.analyseStrict(p.first, p.second)))
                .collect(Collectors.toList());
    }

    private List<Pair<Definition, AnalysisState>> getAnalysisStates(List<Definition> definitions) {
        final DefinitionVisitor definitionVisitor = new DefinitionVisitor(factory, evidenceLimit);
        return definitions.stream().map(def -> Pair.make(def, definitionVisitor.getAnalysisState(def)))
                .collect(Collectors.toList());
    }

    private List<AccessVisitor> collectVisitors(List<Pair<Definition, AnalysisState>> states) {
        return states.stream().map(p -> p.second.visitor).collect(Collectors.toList());
    }

    public AnalyserModel.AnalysisResult internalAnalyseStrict(final Definition definition) {
        final DefinitionVisitor definitionVisitor = new DefinitionVisitor(factory, evidenceLimit);
        final AnalysisState analysisState = definitionVisitor.getAnalysisState(definition);

        try {
            classParser.parse(root, analysisState.visitor);
        } catch (IOException e) {
            throw new AnalyserException(e);
        }

        return StrictAnalyser.analyseStrict(definition, analysisState);
    }

    public List<Pair<Definition, AnalyserModel.LooseAnalysisResult>> analyseLoose(final List<Definition> definitions) {
        if (definitions.isEmpty()) {
            return new ArrayList<>();
        } else if (definitions.size() == 1) {
            return Collections.singletonList(Pair.make(definitions.get(0), internalAnalyseLoose(definitions.get(0))));
        } else {
            return internalAnalyseLoose(definitions);
        }
    }

    private List<Pair<Definition, AnalyserModel.LooseAnalysisResult>> internalAnalyseLoose(
            final List<Definition> definitions) {
        final List<Pair<Definition, AnalysisState>> defVisitors = getAnalysisStates(definitions);
        final AccessVisitor visitor = new CompoundAccessVisitor(collectVisitors(defVisitors));

        try {
            classParser.parse(root, visitor);
        } catch (IOException e) {
            throw new AnalyserException(e);
        }

        return defVisitors.stream().map(p -> Pair.make(p.first, LooseAnalyser.analyseLoose(p.first, p.second)))
                .collect(Collectors.toList());
    }

    private AnalyserModel.LooseAnalysisResult internalAnalyseLoose(final Definition definition) {
        final DefinitionVisitor definitionVisitor = new DefinitionVisitor(factory, evidenceLimit);
        final AnalysisState analysisState = definitionVisitor.getAnalysisState(definition);

        try {
            classParser.parse(root, analysisState.visitor);
        } catch (IOException e) {
            throw new AnalyserException(e);
        }

        return LooseAnalyser.analyseLoose(definition, analysisState);
    }
}

File: core/src/main/scala/com/github/fburato/highwheelmodules/core/analysis/ModuleAnalyser.scala
package com.github.fburato.highwheelmodules.core.analysis

import com.github.fburato.highwheelmodules.core.algorithms.{
  CompoundAccessVisitor,
  ModuleDependenciesGraphBuildingVisitor
}
import com.github.fburato.highwheelmodules.model.analysis.{LOOSE, STRICT}
import com.github.fburato.highwheelmodules.model.classpath.{
  AccessVisitor,
  ClassParser,
  ClasspathRoot
}
import com.github.fburato.highwheelmodules.model.modules._
import com.github.fburato.highwheelmodules.utils.TryUtils._

import java.io.IOException
import scala.util.{Failure, Success, Try}

trait ModuleAnalyser {
  def analyse(definitions: Seq[Definition]): Try[Seq[AnalysisResult]]
}

object ModuleAnalyser {

  private class Implementation(
    classParser: ClassParser,
    classpathRoot: ClasspathRoot,
    evidenceLimit: Option[Int],
    factory: ModuleGraphFactory
  ) extends ModuleAnalyser {
    private val strictAnalyser = StrictAnalyser
    private val looseAnalyser = LooseAnalyser

    override def analyse(definitions: Seq[Definition]): Try[Seq[AnalysisResult]] =
      if (definitions.isEmpty) {
        Success(Seq())
      } else {
        for {
          visitorsAndProcessors <- visitorAndProcessors(definitions)
          (visitor, processors) = visitorsAndProcessors
          _ <- classParser.parse(classpathRoot, visitor) recoverWith { case e: IOException =>
            Failure(AnalyserException(e))
          }
          results <- sequence(processors.map(p => Try(p())))
        } yield results
      }

    private def initialiseState(definition: Definition): Try[AnalysisState] = {
      val other = HWModule.make("(other)", List()).get

      def generateState(modules: Seq[HWModule]): AnalysisState = {
        val specModuleGraph = factory.buildMetricModuleGraph
        definition.modules.foreach(m => specModuleGraph.addModule(m))
        definition.dependencies.foreach(d =>
          specModuleGraph.addDependency(ModuleDependency(d.source, d.dest))
        )
        val actualModuleGraph = factory.buildMetricModuleGraph
        val auxTrackingBareGraph = factory.buildTrackingModuleGraph
        val trackingGraph = factory.buildEvidenceModuleGraph(auxTrackingBareGraph, evidenceLimit)
        val moduleGraphVisitor = ModuleDependenciesGraphBuildingVisitor(
          modules,
          actualModuleGraph,
          other,
          (sourceModule, destModule, _, _, _) => ModuleDependency(sourceModule, destModule),
          definition.whitelist,
          definition.blacklist
        )
        val evidenceGraphVisitor = ModuleDependenciesGraphBuildingVisitor(
          modules,
          trackingGraph,
          other,
          (sourceModule, destModule, sourceAP, destAP, _) =>
            EvidenceModuleDependency(sourceModule, destModule, sourceAP, destAP),
          definition.whitelist,
          definition.blacklist
        )
        val accessVisitor = CompoundAccessVisitor(Seq(moduleGraphVisitor, evidenceGraphVisitor))
        AnalysisState(
          modules,
          definition.dependencies,
          definition.noStrictDependencies,
          specModuleGraph,
          actualModuleGraph,
          auxTrackingBareGraph,
          accessVisitor,
          other
        )
      }

      definition.modules match {
        case Seq() => Failure(AnalyserException("No modules provided in definition"))
        case s     => Success(generateState(s))
      }
    }

    private def visitorAndProcessors(
      definitions: Seq[Definition]
    ): Try[(AccessVisitor, Seq[() => AnalysisResult])] = {
      val merged = definitions.map(d =>
        for {
          state <- initialiseState(d)
        } yield (
          state.visitor,
          () =>
            d.mode match {
              case STRICT => strictAnalyser.analyse(state)
              case LOOSE  => looseAnalyser.analyse(state)
            }
        )
      )
      sequence(merged).map(sequence => {
        val (visitors, processors) = sequence.unzip
        (CompoundAccessVisitor(visitors), processors)
      })
    }
  }

  def apply(
    classParser: ClassParser,
    classpathRoot: ClasspathRoot,
    evidenceLimit: Option[Int],
    factory: ModuleGraphFactory
  ): ModuleAnalyser =
    new Implementation(classParser, classpathRoot, evidenceLimit, factory)
}
